core/filterconstraint.cpp

```diff
@@ -0,0 +1,939 @@
+		return type_descriptions[index].type;
+	return (enum filter_constraint_type)-1;
+}
+
+extern "C"
+enum filter_constraint_string_mode filter_constraint_string_mode_from_index(int index)
+{
+	if (index >= 0 && index < (int)string_mode_descriptions_count())
+		return string_mode_descriptions[index].mode;
+	return (enum filter_constraint_string_mode)-1;
+}
+
+extern "C"
+enum filter_constraint_range_mode filter_constraint_range_mode_from_index(int index)
+{
+	if (index >= 0 && index < (int)range_mode_descriptions_count())
+		return range_mode_descriptions[index].mode;
+	return (enum filter_constraint_range_mode)-1;
+}
+
+QString filter_constraint_type_to_string_translated(enum filter_constraint_type type)
+{
+	const type_description *desc = get_type_description(type);
+	return desc ? gettextFromC::tr(desc->translated) : "unknown";
+}
+
+QString filter_constraint_string_mode_to_string_translated(enum filter_constraint_string_mode type)
+{
+	const string_mode_description *desc = get_string_mode_description(type);
+	return desc ? gettextFromC::tr(desc->translated) : "unknown";
+}
+
+QString filter_constraint_range_mode_to_string_translated(enum filter_constraint_range_mode type)
+{
+	const range_mode_description *desc = get_range_mode_description(type);
+	return desc ? gettextFromC::tr(desc->translated) : "unknown";
+}
+
+QString filter_constraint_negate_to_string_translated(bool negate)
+{
+	return gettextFromC::tr(negate_description[negate]);
+}
+
+static enum filter_constraint_units type_to_unit(enum filter_constraint_type type)
+{
+	const type_description *desc = get_type_description(type);
+	return desc ? desc->units : FILTER_CONSTRAINT_NO_UNIT;
+}
+
+QString filter_constraint_get_unit(enum filter_constraint_type type)
+{
+	switch (type_to_unit(type)) {
+	case FILTER_CONSTRAINT_NO_UNIT:
+	default:
+		return QString();
+	case FILTER_CONSTRAINT_LENGTH_UNIT:
+		return get_depth_unit();
+	case FILTER_CONSTRAINT_TEMPERATURE_UNIT:
+		return get_temp_unit();
+	case FILTER_CONSTRAINT_WEIGHT_UNIT:
+		return get_weight_unit();
+	case FILTER_CONSTRAINT_VOLUMETRIC_FLOW_UNIT:
+		return get_volume_unit();
+	}
+}
+
+static int float_to_int(double f, enum filter_constraint_type type)
+{
+	const type_description *desc = get_type_description(type);
+	if (!desc)
+		return -1;
+	switch (desc->units) {
+	case FILTER_CONSTRAINT_NO_UNIT:
+	default:
+		return (int)lrint(f);
+	case FILTER_CONSTRAINT_LENGTH_UNIT:
+		return prefs.units.length == units::METERS ? lrint(f * 1000.0) : feet_to_mm(f);
+	case FILTER_CONSTRAINT_TEMPERATURE_UNIT:
+		return prefs.units.temperature == units::CELSIUS ? C_to_mkelvin(f) : F_to_mkelvin(f);
+	case FILTER_CONSTRAINT_WEIGHT_UNIT:
+		return prefs.units.weight == units::KG ? lrint(f * 1000.0) : lbs_to_grams(f);
+	case FILTER_CONSTRAINT_VOLUMETRIC_FLOW_UNIT:
+		return prefs.units.volume == units::LITER ? lrint(f * 1000.0) : lrint(cuft_to_l(f) * 1000.0);
+	}
+}
+
+static double int_to_float(int i, enum filter_constraint_type type)
+{
```

Comment:
likeways, `base_to_display_unit()` or something like that.

Reply:
done